cmd/source_profile.go

```diff
@@ -0,0 +1,274 @@
+}
+
+func NewSourceProfileConnection(source string, params map[string]string) (SourceProfileConnection, error) {
+	conn := SourceProfileConnection{}
+	switch strings.ToLower(source) {
+	case "mysql":
+		{
+			conn.ty = SourceProfileConnectionTypeMySQL
+			conn.mysql = NewSourceProfileConnectionMySQL(params)
+		}
+	case "postgresql", "postgres", "pg":
+		{
+			conn.ty = SourceProfileConnectionTypePostgreSQL
+			conn.pg = NewSourceProfileConnectionPostgreSQL(params)
+		}
+	case "dynamodb":
+		{
+			conn.ty = SourceProfileConnectionTypeDynamoDB
+			dydb, err := NewSourceProfileConnectionDynamoDB(params)
+			conn.dydb = dydb
+			if err != nil {
+				return conn, err
+			}
+		}
+	default:
+		return conn, fmt.Errorf("invalid source database: %v", source)
+	}
+	return conn, nil
+}
+
+type SourceProfileConfig struct {
+	path string
+}
+
+func NewSourceProfileConfig(path string) SourceProfileConfig {
+	return SourceProfileConfig{path: path}
+}
+
+type SourceProfile struct {
+	ty     SourceProfileType
+	file   SourceProfileFile
+	conn   SourceProfileConnection
+	config SourceProfileConfig
+}
+
+func (src SourceProfile) ToLegacyDriver(source string) (string, error) {
+	switch src.ty {
+	case SourceProfileTypeFile:
+		{
+			switch strings.ToLower(source) {
+			case "mysql":
+				return "mysqldump", nil
+			case "postgresql", "postgres", "pg":
+				return "pg_dump", nil
+			default:
+				return "", fmt.Errorf("invalid source database %v for source-profile type file", strings.ToLower(source))
+			}
+		}
+	case SourceProfileTypeConnection:
+		{
+			switch strings.ToLower(source) {
+			case "mysql":
+				return "mysql", nil
+			case "postgresql", "postgres", "pg":
+				return "postgres", nil
+			case "dynamodb":
+				return "dynamodb", nil
+			default:
+				return "", fmt.Errorf("invalid source database %v for source-profile direct connect", strings.ToLower(source))
+			}
+		}
+	case SourceProfileTypeConfig:
+		return "", fmt.Errorf("specifying source-profile using config not implemented")
+	default:
+		return "", fmt.Errorf("invalid source-profile, could not infer type")
+	}
+}
+
+// Source profile is passed as a list of key value pairs on the command line.
+// Following 3 formats are supported as valid source profiles.
+//
+// Format 1. Specify file path and file format.
+// File path can be a local file path or a gcs file path. Support for more file
+// path types can be added in future.
+// File format can be "dump" e.g., when specifying a mysqldump or pgdump etc.
+// Support for more formats e.g., "csv", "avro" etc can be added in future.
+//
+// Example: -source-profile="file=/tmp/abc, format=dump"
+// Example: -source-profile="file=gcs://bucket_name/cart.txt, format=dump"
+//
+// Format 2. Specify source connection parameters. If none specified, then read
+// from envrironment variables.
+//
+// Format 3. Specify a config file that specifies source connection profile.
+//
+func NewSourceProfile(str string, source string) (SourceProfile, error) {
```

Comment:
Suggest changing "str" arg name to either just "s" or something more descriptive e.g. "profile".
Just "str" isn't very useful.

Reply:
I think just "s" works here, since the scope for this argument is pretty limited. Done.